b2sdk/bucket.py

```diff
@@ -494,194 +400,127 @@ def upload(
-            # so that they share resources during a sync.
-            f = self.api.get_thread_pool().submit(
-                self._upload_small_file, upload_source, file_name, content_type, file_info,
-                progress_listener
+        if recomended_part_size is not None:
+            planner = EmergePlanner.from_account_info(
+                self.api.session.account_info,
+                recomended_part_size=recomended_part_size
             )
-            return f.result()
         else:
-            return self._upload_large_file(
-                upload_source, file_name, content_type, file_info, progress_listener
-            )
-
-    def _upload_small_file(
-        self, upload_source, file_name, content_type, file_info, progress_listener
-    ):
-        content_length = upload_source.get_content_length()
-        exception_info_list = []
-        progress_listener.set_total_bytes(content_length)
-        with progress_listener:
-            for _ in six.moves.xrange(self.MAX_UPLOAD_ATTEMPTS):
-                try:
-                    with upload_source.open() as file:
-                        input_stream = ReadingStreamWithProgress(file, progress_listener)
-                        hashing_stream = StreamWithHash(input_stream)
-                        length_with_hash = content_length + hashing_stream.hash_size()
-                        response = self.api.session.upload_file(
-                            self.id_, file_name, length_with_hash, content_type, HEX_DIGITS_AT_END,
-                            file_info, hashing_stream
-                        )
-                        assert hashing_stream.hash == response['contentSha1']
-                        return FileVersionInfoFactory.from_api_response(response)
-                except B2Error as e:
-                    if not e.should_retry_upload():
-                        raise
-                    exception_info_list.append(e)
-                    self.api.account_info.clear_bucket_upload_data(self.id_)
-
-        raise MaxRetriesExceeded(self.MAX_UPLOAD_ATTEMPTS, exception_info_list)
-
-    def _upload_large_file(
-        self, upload_source, file_name, content_type, file_info, progress_listener
-    ):
-        content_length = upload_source.get_content_length()
-        if self.MAX_LARGE_FILE_SIZE < content_length:
-            raise MaxFileSizeExceeded(content_length, self.MAX_LARGE_FILE_SIZE)
-        minimum_part_size = self.api.account_info.get_minimum_part_size()
-
-        # Set up the progress reporting for the parts
-        progress_listener.set_total_bytes(content_length)
-
-        # Select the part boundaries
-        part_ranges = choose_part_ranges(content_length, minimum_part_size)
-
-        # Check for unfinished files with same name
-        unfinished_file, finished_parts = self._find_unfinished_file_if_possible(
-            upload_source,
+            planner = None
+        return emerger_method(
```

Comment:
problem is that `absoluteMinimumPartSize` and others params are available only after authorizing - I got an error when planner was created with Emerger - so now it is created dynamically - I wanted for `min_part_size` `recommended_part_size` and `max_part_size` to be params of planner - not params of `get_emerge_plan` method - also I was thinking that maybe one day we would want different planner for streaming version non-streaming version - also, maybe for some future advanced usage some other planner would be needed...

but I am not too attached to this - I don't know why would I need AbstractEmerger - but we can easily just pass `recommended_part_size` to Emerger and let it contruct planner...

Reply:
maybe you can get `absoluteMinimumPartSize` etc dynamically (maybe a getter that just gets it from somewhere else whenever you'd like to access the value) instead of copying the value into the class + mnemonization... nah, that's ugly too.

I don't really have a good idea yet on how to structure it so that it is elegant, but I'm pretty sure we can do significantly better than the current solution. Lets try with what you wrote in the last paragraph, maybe it will be enogh.